File: lib/android/vesper-player-kit/src/main/java/io/github/umbrella22/vesper/player/android/VesperVideoPresentation.kt
package io.github.umbrella22.vesper.player.android

/** Display dimensions after native rotation and pixel aspect correction. */
data class VesperVideoPresentation(
    val displayWidth: Double,
    val displayHeight: Double,
) {
    val displayAspectRatio: Double get() = displayWidth / displayHeight
}

/** Attached surface geometry in dp, relative to the host's top-left corner. */
data class VesperVideoSurfaceGeometry(
    val width: Double,
    val height: Double,
    val contentRect: VesperVideoRect,
)

data class VesperVideoRect(
    val left: Double,
    val top: Double,
    val width: Double,
    val height: Double,
)

internal fun NativeVideoLayoutInfo.toVideoPresentation(): VesperVideoPresentation? {
    if (width <= 0 || height <= 0 || !pixelWidthHeightRatio.isFinite() || pixelWidthHeightRatio <= 0f) return null
    return VesperVideoPresentation(width.toDouble() * pixelWidthHeightRatio, height.toDouble())
}

File: lib/ios/VesperPlayerKit/Sources/VesperPlayerKit/Playback/VesperVideoPresentation.swift
import Combine
import CoreGraphics

/// Display dimensions after native rotation and pixel aspect correction.
public struct VesperVideoPresentation: Equatable {
    public let displayWidth: Double
    public let displayHeight: Double
    public var displayAspectRatio: Double { displayWidth / displayHeight }

    init?(size: CGSize) {
        guard size.width.isFinite, size.height.isFinite, size.width > 0, size.height > 0 else { return nil }
        displayWidth = Double(size.width)
        displayHeight = Double(size.height)
    }
}

/// Geometry in local points for one attached playback view.
public struct VesperVideoSurfaceGeometry: Equatable {
    public let width: Double
    public let height: Double
    public let contentRect: CGRect
}

@MainActor
final class VesperVideoPresentationState {
    @Published private(set) var value: VesperVideoPresentation?

    func update(_ value: VesperVideoPresentation?) {
        if self.value != value { self.value = value }
    }
}
